xapian-bindings/perl/perl.i

```diff
@@ -405,12 +405,45 @@ sub add_rangeprocessor {
 }
 %}
 
+%perlcode %{
+package Xapian::FieldProcessor::AutomaticCodeWrapper {
+    our @ISA = qw(Xapian::FieldProcessor);
+
+    sub new {
+        my $class = shift;
+        my $coderef = shift;
+
+        my $self = $class->SUPER::new(@_);
+
+        $self{_fpcoderef} = $coderef;
+        return $self;
+    }
+
+    sub __call__ {
+        my ($self, $s) = @_;
+
+        my $res = $self{_fpcoderef}->($s);
+        # $res will pass the boundary to C++-land, so we need to make sure
+        # it persists...
+        push @{$self{_calledresrefs}}, $res;
```

Comment:
This means that any value returned by the FieldProcessor can't be reclaimed while the object which returned it lives, which isn't ideal - consider someone parsing a billion queries with the same object.

I wonder if it would be better to implement this like the old hand-coded XS bindings did for ExpandDecider by using a C++ subclass, which I think avoids this problem: https://git.xapian.org/?p=xapian;a=blob;f=search-xapian/Xapian.xs;h=c6effe5049fe4fc88047b42b2d6a74469da4b7fa;hb=refs/heads/svn/1.2#l202

Reply:
Yup. Better leaking memory than violating one's segmentations, but I can see your point :) I will see what I can do, but I think I won't be able to do so before next weekend.